docs/design_proposals/event-api-v2.md

```diff
@@ -0,0 +1,119 @@
+
+To address this, **I propose that we change the Build() interface to return a list of build results, one for each artifact, with information about the original artifact skaffold attempted to build, and either the built artifact, or the error the builder encountered.**
+
+This will allow the runner to:
+* know which artifacts were built successfully, and which failed
+* log events for each artifact, complete with a detailed error message passed back from the runner in the event that the build didn't complete successfully
+
+With this information, we could move all of the event handling code into a decorator that we wrap the builders and deployers in during the creation of the runner, similar to the way we handle timings currently. The semantics for each state (Not Started, In Progress, Complete, Failed) and the events that transition the states should all be the same between builders and deployers, and we can share the logic for handling each of these events in a separate location. This has the added bonus of not requiring any event-related code to be written by contributors if/when they add new builders/deployers, or make changes to existing ones.
+
+
+## Implementation plan
+
+Implementing this should be pretty straightforward:
+
+1) **Rework the Build() interface**
+
+The current method signature is 
+
+```golang
+Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+```
+
+This will be changed to return a list of build results for each artifact, along with the artifact information itself. The artifact information can be embedded in the build result:
+
+```golang
+type BuildResult struct {
+  Target *latest.Artifact
+  Result *build.Artifact
+  Error error
+}
+```
+and the new method signature will be
+```golang
+Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]BuildResult, error)
+```
+
+2) **Consume the new build results in the runner**
+
+The runner can then take control of all the eventing code. Deploy results are already returned by the Deploy() interface implementations, so with the Build results we have all the information we need. Using Build() as an example (Deploy() will look similar):
+
+* For each artifact provided to Build(), an `BuildInProgress` event is triggered
+* Once the build is complete, for each returned result, trigger a `BuildFailed` event for results with errors, and a `BuildComplete` event for those without.
+
+```golang
+for _, a := range artifactsToBuild {
+  event.BuildInProgress(a)
+}
+bRes, err := r.Build(ctx, out, tags, artifactsToBuild)
+for _, res := range bRes {
+  if res.Error != nil {
+    event.BuildFailed(res.Target, res.Error)
+  } else {
+    event.BuildComplete(res.Target)
```

Comment:
Who are consuming these events?
Is it the IDE team? If that is the case, we should get them involved as a stake holder.

Reply:
In general I agree, and they are actively testing the functionality.
This redesign is targeting the internal details and should not impact the functionality of the Event API. Not sure if it would be productive to involve them at this level, they are interfacing with the HTTP/gRPC level.

Reply:
Well, in this re-design we are changing the functionality a little bit since we are changing when the events get triggered.

Reply:
Actually - I take that back - the change in "streaming" vs "batch" events for multi-artifact builds is very much worthwhile to bring up to the IDE teams.